### DISCUSSION TITLE
How to override database URL with test database URL when testing?

### DISCUSSION THREAD

**Discussion by masahiro04** [fastapi-users/fastapi-users#523] [Q&A]
I can test with the development database, but is there a way to override the database URL with "test database URL" with the AsyncClient as in the official docs with the fastapi TestClient?

The fastapi test docs have an article about applying Session Storage, It is possible to deal with it by overriding the test database as it is, but there is no article when using AsyncClient.
[Testing a Database](https://fastapi.tiangolo.com/advanced/testing-database/)

```
# TestClient -> Works.
SQLALCHEMY_DATABASE_URL = "sqlite:///./test.db"
engine = create_engine(
    SQLALCHEMY_DATABASE_URL, connect_args={"check_same_thread": False}
)
TestingSessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
Base.metadata.create_all(bind=engine)

def override_get_db():
    try:
        db = TestingSessionLocal()
        yield db
    finally:
        db.close()

app.dependency_overrides[get_db] = override_get_db
client = TestClient(app)

#-----------

# AsyncClient -> How to override database URL?
@pytest.mark.asyncio
import httpx
from asgi_lifespan import LifespanManager

async with LifespanManager(app):
    async with AsyncClient(app=app, base_url=base_url) as client:
        ...
```
Any advices would be helpful.
Thanks.

**Comment by frankie567** [comment]
Hi there!

`dependency_overrides` is a feature of FastAPI, so you can use it exactly the same way:

```py
app.dependency_overrides[get_db] = override_get_db
async with LifespanManager(app):
    async with AsyncClient(app=app, base_url=base_url) as client:
```

**Reply by masahiro04** [reply]
@frankie567 
I confirmed by debugging that if I override get_db and database, it will be a changed from development database to test database.


![Screen Shot 2021-02-22 at 16 55 11](https://user-images.githubusercontent.com/23392138/108678699-ca6cb780-752e-11eb-8c96-01d7770a516b.png)

However, I think there is a problem because the data is stored in the development database instead of the test database. Is there any other setting required?




```py
"""
database.py
"""

SQLALCHEMY_DATABASE_URL = os.getenv("DATABASE_URL")

engine = create_engine(SQLALCHEMY_DATABASE_URL)

# fastapi-users のために設定
database = databases.Database(SQLALCHEMY_DATABASE_URL)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

Base: DeclarativeMeta = declarative_base()

"""
dependencies.py
"""

users = User.__table__
user_db = SQLAlchemyUserDatabase(UserDB, database, users)

SECRET = "SECRET"

auth_backends = []

jwt_authentication = JWTAuthentication(
    secret=SECRET, lifetime_seconds=36000, tokenUrl="/api/v1/auth/jwt/login"
)

auth_backends.append(jwt_authentication)

fastapi_users = FastAPIUsers(
    user_db,
    auth_backends,
    sUser,
    UserCreate,
    UserUpdate,
    UserDB,
)


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


def get_current_user():
    return fastapi_users.current_user()


"""
conftest.py
"""
TEST_SQLALCHEMY_DATABASE_URL = "sqlite:///./test_temp.db"
nest_asyncio.apply()


@pytest.fixture(scope="session")
def SessionLocal():
    # settings of test database
    engine = create_engine(TEST_SQLALCHEMY_DATABASE_URL, connect_args={"check_same_thread": False})
    assert not database_exists(
        TEST_SQLALCHEMY_DATABASE_URL), "Test database already exists. Aborting tests."

    # Create test database and tables
    Base.metadata.create_all(engine)
    SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

    # Run the tests
    yield SessionLocal

    # Drop the test database
    drop_database(TEST_SQLALCHEMY_DATABASE_URL)


base_url = "http://0.0.0.0"


@pytest.fixture(scope="module")
async def async_client(SessionLocal) -> Generator:
    app.dependency_overrides[get_db] = SessionLocal()

    # override database URL with test database URL
    test_database = databases.Database(TEST_SQLALCHEMY_DATABASE_URL)
    app.dependency_overrides[database] = test_database
    async with LifespanManager(app):
        async with AsyncClient(app=app, base_url=base_url) as client:
            yield client


```

**Comment by masahiro04** [comment]
@frankie567 

I've checked and I understand that I should override get_current_user in dependencies.
Sorry for my immature understanding of fastapi.

---

I want to override `get_current_user` with mock data  in my test case. Is there a data generation class for testing?
Or can it be created with `fastapi_users.create()`?

**Reply by frankie567** [reply]
Currently, there is no built-in tool to help testing ; but this is something I have in mind.

I think you can create a test user with `fastapi_users.create()` and use it to override `get_current_user`, yes.

**Comment by DiEVeXx** [comment]
I am having this same problem, I would like to be able to store and get the users from the test database instead of the development database.

I am trying to do a dependency_overrides on the get_users_db function that the user manager needs, but I can't get this to work.

Any news or way to do it?

**Reply by frankie567** [reply]
You should override the `get_async_session` dependency to return a session targetting the test database. This way, all your FastAPI tests will run against the test database.
